### PR TITLE
Feat(users): Prevent deletion of external users.

### PR DESCRIPTION
**Pull request by psantos1113**
Creating a Draft pull request while I sort out what is wrong with SAML authentication.

### CODE REVIEW

File: alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java [hunk 1]
```diff
@@ -149,25 +158,36 @@ public boolean changeUserEmailAddress(String username, String emailAddress) {
...
     @Override
     public void deleteUser(String userName) throws AlertForbiddenOperationException {
-        Optional<Long> optionalUserId = userRepository.findByUserName(userName).map(UserEntity::getId);
-        if (optionalUserId.isPresent()) {
-            deleteUserById(optionalUserId.get());
+        Optional<UserEntity> optionalUser = userRepository.findByUserName(userName);
+        if (optionalUser.isPresent()) {
+            UserEntity user = optionalUser.get();
+            if (user.isExternal()) {
+                throw new AlertForbiddenOperationException(String.format("The '%s' cannot be deleted", user.getUserName()));
```
Comment (gkillough) [inline alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java]: You might want to add `user` after the username in the string.

File: alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java [hunk 2]
```diff
@@ -149,25 +158,36 @@ public boolean changeUserEmailAddress(String username, String emailAddress) {
...
     }
 
     @Override
     public void deleteUser(Long userId) throws AlertForbiddenOperationException {
-        deleteUserById(userId);
+        Optional<UserEntity> optionalUser = userRepository.findById(userId);
+        if (optionalUser.isPresent()) {
+            UserEntity user = optionalUser.get();
+            if (user.isExternal()) {
+                throw new AlertForbiddenOperationException(String.format("The '%s' cannot be deleted", user.getUserName()));
```
Comment (gkillough) [inline alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java]: `user`

File: alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java [hunk 3]
```diff
@@ -149,25 +158,36 @@ public boolean changeUserEmailAddress(String username, String emailAddress) {
 
     @Override
     public void deleteUser(String userName) throws AlertForbiddenOperationException {
-        Optional<Long> optionalUserId = userRepository.findByUserName(userName).map(UserEntity::getId);
-        if (optionalUserId.isPresent()) {
-            deleteUserById(optionalUserId.get());
+        Optional<UserEntity> optionalUser = userRepository.findByUserName(userName);
+        if (optionalUser.isPresent()) {
```
Comment (bamandel) [inline alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultUserAccessor.java]: This code in the if statement and beyond is identical to the other delete method. We can probably abstract some of this to another method.

File: src/main/resources/db/changelog/alert/5.3.0/changelog-user-management.xml [hunk 4]
```diff
@@ -10,4 +10,11 @@
             CALL REMOVE_ALL_PERMISSIONS('ALERT_USER','component_authentication', 'GLOBAL');
         </sql>
     </changeSet>
+    <changeSet author="psantos" id="2019-12-18-07-31-16-844">
+        <addColumn schemaName="ALERT" tableName="USERS">
+            <column name="EXTERNAL" type="BOOLEAN" defaultValueBoolean="false">
```
Comment (bamandel) [inline src/main/resources/db/changelog/alert/5.3.0/changelog-user-management.xml]: Won't all the users (Except the ones we created) be external at this point?

Comment (psantos1113) [reply src/main/resources/db/changelog/alert/5.3.0/changelog-user-management.xml]: I will add the column to 5.2.0 therefore this change can be omitted from this PR.

Comment (psantos1113) [reply src/main/resources/db/changelog/alert/5.3.0/changelog-user-management.xml]: In 5.2.0 the send of the authentication event is commented out.  So only in 5.3.0 will we sync external users.  Therefore this changeset is correct for 5.3.0 only

**Pull request merged**